deflate_stored.c

```diff
@@ -0,0 +1,194 @@
+/* deflate_stored.c -- store data without compression using deflation algorithm
+ *
+ * Copyright (C) 1995-2013 Jean-loup Gailly and Mark Adler
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+#include "zbuild.h"
+#include "deflate.h"
+#include "deflate_p.h"
+#include "functable.h"
+
+unsigned read_buf(PREFIX3(stream) *strm, unsigned char *buf, unsigned size);
+
+/* ===========================================================================
+ * Copy without compression as much as possible from the input stream, return
+ * the current block state.
+ *
+ * In case deflateParams() is used to later switch to a non-zero compression
+ * level, s->matches (otherwise unused when storing) keeps track of the number
+ * of hash table slides to perform. If s->matches is 1, then one hash table
+ * slide will be done when switching. If s->matches is 2, the maximum value
+ * allowed here, then the hash table will be cleared, since two or more slides
+ * is the same as a clear.
+ *
+ * deflate_stored() is written to minimize the number of times an input byte is
+ * copied. It is most efficient with large input and output buffers, which
+ * maximizes the opportunites to have a single copy from next_in to next_out.
+ */
+Z_INTERNAL block_state deflate_stored(deflate_state *s, int flush) {
+    /* Smallest worthy block size when not flushing or finishing. By default
+     * this is 32K. This can be as small as 507 bytes for memLevel == 1. For
+     * large input and output buffers, the stored block size will be larger.
+     */
+    unsigned min_block = MIN(s->pending_buf_size - 5, s->w_size);
+
+    /* Copy as many min_block or larger stored blocks directly to next_out as
+     * possible. If flushing, copy the remaining available input to next_out as
+     * stored blocks, if there is enough space.
+     */
+    unsigned len, left, have, last = 0;
+    unsigned used = s->strm->avail_in;
+    do {
+        /* Set len to the maximum size block that we can copy directly with the
+         * available input data and output space. Set left to how much of that
+         * would be copied from what's left in the window.
+         */
+        len = MAX_STORED;       /* maximum deflate stored block length */
+        have = (s->bi_valid + 42) >> 3;         /* number of header bytes */
+        if (s->strm->avail_out < have)          /* need room for header */
+            break;
+            /* maximum stored block length that will fit in avail_out: */
+        have = s->strm->avail_out - have;
+        left = (int)s->strstart - s->block_start;    /* bytes left in window */
+        if (len > (unsigned long)left + s->strm->avail_in)
+            len = left + s->strm->avail_in;     /* limit len to the input */
```

Comment:
Both `len` and `left` are defined as `unsigned int`, so when `left` is cast to larger type, the comparison wraps around as `len` can't be larger than what `unsigned int` can fit. 

Obviously `len` on line 54 might be implicitly zero-extended to `unsigned long`, but that makes the comparison (almost) always false.

Reply:
I'm not sure why we can't use `MIN()` here.

Reply:
MIN() as a macro only works with variables, not expressions... This is limitation of preprocessor. 

Instead of relying on implicit cast, which is non-standard (standard says cast to `int` if types are different), it should have explicit cast on both sides of comparison.

Reply:
> MIN() as a macro only works with variables, not expressions... This is limitation of preprocessor.

`len = MIN(len, left + s->strm->avail_in);` compiles fine.

Reply:
It does compile, but the result of second parameter can overflow... That is why the original code had cast to `unsigned long`...

Problem of the original code was that it assumed `len` was automatically zero-extended to `unsigned long` before comparison... Some compilers do it, but that is not standard behavior. Essentially it creates temporary variable that is discarded right after the comparison. I would say it's waste of registers as zero-extension can't be done for variable stored in memory.

Reply:
Ok.